rfd/00XY-session-locking.md

```diff
@@ -0,0 +1,269 @@
+automatically after a failure.  If a tolerance interval is exceeded while performing
+such reloads, the fallback mode described above is employed until a healthy connection
+is reestablished.
+
+The developed logic should work with every protocol that makes use of
+`srv.Monitor`: SSH, k8s and DB.
+
+### Replication to leaf clusters
+
+`SessionLock` resources are replicated from the root cluster to leaf clusters
+in a similar manner to how CAs are shared between trusted clusters.
+
+The goal should be achieved by introducing a routine similar to
+`periodicUpdateCertAuthorities`. However instead of polling the backend (with
+the default period of 10 minutes defined in `defaults.LowResPollingPeriod`)
+a `types.Watcher`-based algorithm should be preferred.
+
+To be able to distinguish session locks received from the root cluster,
+the root cluster should send the `SessionLock` resource a resource label of the
+form:
+```
+teleport.dev/replicated-from: <root-cluster-name>
+```
+
+A tolerance interval similar to the case of intra-cluster propagation should be used.
+When a leaf cluster's reverse tunnel connection exhibits intolerable failures,
+the leaf cluster's fallback mode will be enforced with respect to the connections
+authenticated with a user certificate issued by the root cluster's CA.
+
+### Scenarios
+
+#### Creating a permanent lock
+
+```
+$ tctl sessions lock --user=foo@example.com --message="Suspicious activity."
+Created a session lock with ID "dc7cee9d-fe5e-4534-a90d-db770f0234a1".
+```
+
+This locks out `foo@example.com` without automatic expiration.
+The lock can be lifted by `tctl rm lock/dc7cee9d-fe5e-4534-a90d-db770f0234a1`.
+
+The above locking command would be equivalent to
+```sh
+tctl create <<EOF
+kind: session_lock
+metadata:
+  name: dc7cee9d-fe5e-4534-a90d-db770f0234a1
+spec:
+  message: "Suspicious activity."
+  target:
+    user: foo@example.com
+version: v2
+EOF
+```
+
+The showed YAML would also correspond to the output of `tctl get lock/dc7cee9d-fe5e-4534-a90d-db770f0234a1`.
+
+#### Creating a lock with expiry
+
+```
+$ tctl sessions lock --role=developers --message="Cluster maintenance." --expires-in=10h
+Created a session lock with ID "dc7cee9d-fe5e-4534-a90d-db770f0234a1".
+```
+
+This locks out users with the role `developers` for the next 10 hours.
+
+Assuming the time at which the command is issued on 2021-06-14 at 12:27 UTC,
+the above locking command would be equivalent to both
+```sh
+tctl create <<EOF
+kind: session_lock
+metadata:
+  name: dc7cee9d-fe5e-4534-a90d-db770f0234a1
+spec:
+  target:
+    role: developers
+  message: "Cluster maintenance."
+  expires: "2021-06-14T22:27:00Z"   # RFC3339
+version: v2
+EOF
+```
+and
+```sh
+tctl sessions lock --role=developers --message="Cluster maintenance." --expires="2021-06-14T22:27:00Z"
+```
+
+#### Generation of new user certificates prevented
+
+Attempts to generate a new user certificate (`tsh login`, `tctl auth sign`)
```

Comment:
Not sure if `tctl auth sign` too should be blocked here.

Reply:
It should be blocked for any matching role - that is why `session` is not the best name here, strictly speaking it's locking identity